### PR TITLE
Change QPDF::copyForeignObject to return a null object when called wi…

### LINKED ISSUE #1011: qpdf/libqpdf/QPDFObjectHandle.cc:235:null pointer
When using the qpdf --empty --pages command to merge 1.pdf with other PDF files, a segmentation fault occurred due to a null pointer.

version:11.5.0
My system OS：Ubuntu 22.04
My Compilation Process：
git clone https://github.com/qpdf/qpdf
cd qpdf
cmake -S . -B build -DCMAKE_BUILD_TYPE=RelWithDebInfo ..
cd build
cmake --build .
sudo cmake --install .

reproduce: qpdf --empty --pages 1.pdf 2.pdf -- output.pdf

The stack information is as follows：

Program received signal SIGSEGV, Segmentation fault.
QPDFObjectHandle::disconnect (this=this@entry=0x7fffffffdd20) at /home/ljh/qpdf/libqpdf/QPDFObjectHandle.cc:235
235	        this->obj->disconnect();
(gdb) bt
#0  QPDFObjectHandle::disconnect (this=this@entry=0x7fffffffdd20)
    at /home/ljh/qpdf/libqpdf/QPDFObjectHandle.cc:235
#1  0x00007ffff7dc58f6 in QPDFObjectHandle::DisconnectAccess::disconnect (
    o=...) at /home/ljh/qpdf/include/qpdf/QPDFObjectHandle.hh:1268
#2  QPDF_Dictionary::disconnect (this=0x555555619ba0)
    at /home/ljh/qpdf/libqpdf/QPDF_Dictionary.cc:52
#3  0x00007ffff7cb9056 in QPDFObject::disconnect (this=0x555555617790)
    at /home/ljh/qpdf/libqpdf/qpdf/QPDFObject_private.hh:148
#4  QPDF::~QPDF (this=this@entry=0x5555555a4700, __in_chrg=<optimized out>)
    at /home/ljh/qpdf/libqpdf/QPDF.cc:236
#5  0x00007ffff7c6b4cb in std::default_delete<QPDF>::operator() (
    __ptr=0x5555555a4700, this=<optimized out>)
    at /usr/include/c++/11/bits/unique_ptr.h:85
#6  0x00007ffff7c6f03e in std::default_delete<QPDF>::operator() (
    __ptr=<optimized out>, this=<optimized out>)
    at /usr/include/c++/11/bits/unique_ptr.h:79
#7  std::unique_ptr<QPDF, std::default_delete<QPDF> >::~unique_ptr (
    this=0x7fffffffddc8, __in_chrg=<optimized out>)
    at /usr/include/c++/11/bits/unique_ptr.h:361
#8  QPDFJob::createQPDF (this=<optimized out>)
    at /home/ljh/qpdf/libqpdf/QPDFJob.cc:481
#9  0x00007ffff7d375f7 in QPDFJob::run (this=this@entry=0x7fffffffde70)
    at /home/ljh/qpdf/libqpdf/QPDFJob.cc:516

[1.pdf](https://github.com/qpdf/qpdf/files/12141225/1.pdf)
[2.pdf](https://github.com/qpdf/qpdf/files/12141226/2.pdf)

### PR DESCRIPTION
**Pull request by m-holger**
…th a /Pages object (fixes #1011)

If copyForeignObject is called with a pages object, the object will not be inserted into the obj_copier.obj_map during normal processing. In the return statement the missing entry is default constructed, with the consequence that an uninitialized object handle is returned. All future references to the object are also replaced with an uninitialized object rather than a null object.

### CODE REVIEW

**Comment by m-holger** [comment]
There probably should be a test case for this.

**Comment by jberkenbilt** [comment]
Is this made obsolete by #1028?

**Comment by m-holger** [comment]
No. This one deals with severely damaged files with garbage objects pretending to be page tree nodes

**Comment by jberkenbilt** [comment]
I agree that there should be a test for this. Do we know how to reproduce it? Anyway, the new code is clearly better than the old code, so I'm inclined to merge it, though I'm worried that taking the change will cover up some other bug. I think I need to understand better when this happens. I think it would be better if we never inserted uninitialized objects rather than detecting them and returning null instead.

**Comment by m-holger** [comment]
> Do we know how to reproduce it? 

All that is required is to change the /Type entry of a top level foreign object, e.g. the /AcroForm dictionary, to /Pages. 
That will result in the ObjGen not being inserted into obj_copier.object_map. On trying to retrieve the missing object from the object_map it gets default constructed as an uninitialized object handle. Similarly, deliberately trying to foreign copy a /Pages  object will have the same result.

**Comment by m-holger** [comment]
> Is this made obsolete by https://github.com/qpdf/qpdf/pull/1028?

> No. This one deals with severely damaged files with garbage objects pretending to be page tree nodes

Sorry, it's been a while since I did this. This PR is actually the first commit in #1028

**Comment by m-holger** [comment]
I've added a test.

>  I think it would be better if we never inserted uninitialized objects rather than detecting them and returning null instead.

I thought it was alright given that the check/fix was at the point of default creation but I've updated to move the check to before the map access.

File: libqpdf/QPDF.cc [hunk 1]
```diff
@@ -2034,13 +2034,13 @@ QPDF::copyForeignObject(QPDFObjectHandle foreign)
     }
     obj_copier.to_copy.clear();
 
-    auto& result = obj_copier.object_map[foreign.getObjGen()];
-    if (!result.isInitialized()) {
-        result = QPDFObjectHandle::newNull();
+    auto og = foreign.getObjGen();
+    if (!obj_copier.object_map.count(og)) {
         warn(damagedPDF("Unexpected reference to /Pages object while copying foreign object. "
```
Comment (jberkenbilt) [inline libqpdf/QPDF.cc:2039]: I missed this before, but to be consistent with the style of most warnings, you could you change this to lower-case and semicolon-separated, as in `unexpected reference to /Pages object while copying foreign object;  replacing with null`

**Review by jberkenbilt — CHANGES_REQUESTED** [review]
Great, thanks. Requesting a small change to the warning that I overlooked in my previous review.

**Comment by jberkenbilt** [comment]
I'm stepping away for a few hours, but if this is not updated when I return, I will merge as is and tweak the warning message in a separate commit.

**Review by jberkenbilt — APPROVED** [review]

**Pull request merged**